### PR TITLE
New gapic

### PR DESCRIPTION
**Pull request by stephenplusplus**
Blocked by: https://github.com/googleapis/toolkit/issues/1623

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -72,11 +72,27 @@ function Logging(options) {
     return new Logging(options);
   }
 
+  // Determine what scopes are needed.
+  // It is the union of the scopes on all three clients.
+  let scopes = [];
```
Comment (stephenplusplus) [inline src/index.js]: @lukesneeringer It doesn't seem the burden of this code should be on the client library. It looks like we're asking the generated library for information only to give it back.

Comment (lukesneeringer) [reply src/index.js]: I probably agree but do not want to block on it.

**Comment by stephenplusplus** [comment]
#### Unit tests

`npm test` is being a little crazy for me:

```
  1) ConfigServiceV2Client
       getExclusion
         invokes getExclusion without error:
     TypeError: client.exclusionPath is not a function
      at Context.done (C:\Users\sawch\dev\nodejs-logging/test\gapic-v2.js:654:34)

  2) ConfigServiceV2Client
       getExclusion
         invokes getExclusion with error:
     TypeError: client.exclusionPath is not a function
      at Context.done (C:\Users\sawch\dev\nodejs-logging/test\gapic-v2.js:688:34)

  3) ConfigServiceV2Client
       updateExclusion
         invokes updateExclusion without error:
     TypeError: client.exclusionPath is not a function
      at Context.done (C:\Users\sawch\dev\nodejs-logging/test\gapic-v2.js:778:34)

  4) ConfigServiceV2Client
       updateExclusion
         invokes updateExclusion with error:
     TypeError: client.exclusionPath is not a function
      at Context.done (C:\Users\sawch\dev\nodejs-logging/test\gapic-v2.js:816:34)

  5) ConfigServiceV2Client
       deleteExclusion
         invokes deleteExclusion without error:
     TypeError: client.exclusionPath is not a function
      at Context.done (C:\Users\sawch\dev\nodejs-logging/test\gapic-v2.js:846:34)

  6) ConfigServiceV2Client
       deleteExclusion
         invokes deleteExclusion with error:
     TypeError: client.exclusionPath is not a function
      at Context.done (C:\Users\sawch\dev\nodejs-logging/test\gapic-v2.js:864:34)
```

#### Excessive output

```
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 191): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 192): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 193): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 194): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 195): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 196): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 197): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 198): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 199): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 200): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 201): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 202): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 203): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 204): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 205): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 206): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 207): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 208): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 209): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
(node:9756) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 210): Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
    updateExclusion
      3) invokes updateExclusion without error
```

That line `Could not load...` prints out hundreds of times within the test run, only during the gapic tests. They go away when I use ADC.

#### System tests

`npm run system-test` has a new error:

```
      1) should write a single entry to a log
(node:4076) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: Cannot read property 'length' of undefined
```

**Comment by lukesneeringer** [comment]
CI is getting the same thing you are. @landrito is looking into it.

**Comment by landrito** [comment]
The first error is fixed in the most recent commit. 
The second was something that was reported on gcn that we were never able to look into.
 - essentially we are not handling not finding credentials correctly.
The third error is still a bundling issue: It appears when we migrated over to protobufjs6, they changed the return type of the message.encode function. I have a fix and validated that it fixed the system test. I'll have a PR for that fix by the morning.

**Comment by landrito** [comment]
Bah actually I need to a bit of digging. Looks like the fixing the bytelengthfunction in batching didn't cause the system test to pass. Now the next test is getting hung.

**Comment by lukesneeringer** [comment]
> The third error is still a bundling issue: It appears when we migrated over to protobufjs6, they changed the return type of the message.encode function. I have a fix and validated that it fixed the system test. I'll have a PR for that fix by the morning.

I am fairly certain I fixed this already. `message.encode(thing).len` works.

**Comment by lukesneeringer** [comment]
> The second was something that was reported on gcn that we were never able to look into.

I figured this out. Toolkit PR incoming.

**Comment by stephenplusplus** [comment]
I rm -rf'd node_modules && npm cache clean'd and I get more errors in the system tests:

```
  Logging
    sinks
      1) should create a sink with a Bucket destination
      2) should create a sink with a Dataset destination
      3) should create a sink with a Topic destination
      metadata
        4) "before all" hook for "should create a sink with a Topic destination"
      listing sinks
        5) "before all" hook
    logs
      √ should list log entries (590ms)
      √ should list log entries as a stream (223ms)
      6) should write a single entry to a log
(node:15100) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: Cannot read property 'length' of undefined
(node:15100) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
run: Success! Test finished in 6.77s
```

The noise is gone from `npm test`.

**Comment by lukesneeringer** [comment]
@stephenplusplus I assert this should be in good shape now. Please verify. :-)

**Comment by lukesneeringer** [comment]
The existing code breaks the docs build; I will figure out the correct syntax and update in toolkit.

**Comment by lukesneeringer** [comment]
Okay I think this is ready to go. @stephenplusplus please confirm.

**Comment by stephenplusplus** [comment]
Yep, all works for me now!

**Comment by lukesneeringer** [comment]
Can I have a green checkmark, please? ;-)

**Pull request merged**